### DISCUSSION TITLE
SoundDevice: Class Compliant USB Audio Interface

### DISCUSSION THREAD

**Discussion by Koshdukai** [probonopd/MiniDexed#148] [Q&A] [answered]
Hi :)

Any plans to support Class Compliant USB Audio Interfaces as a 4th SoundDevice option in the future?

There's already USB HID (PC Keyb) support and USB MIDI so I don't know (yet) how hard is it to also support USB Audio.

**Comment by probonopd** [comment]
Hello @Koshdukai. Please see https://github.com/rsta2/circle/issues/222. So, probably not anytime soon.

**Reply by probonopd** [reply]
Here is an alternative:

https://github.com/probonopd/MiniDexed/wiki/Hardware#hdmi-to-audio

**Reply by probonopd** [reply]
It looks like @BenZonneveld has implemented "i2s to USB audio":

https://github.com/BenZonneveld/MiniDexed_Midi_Master/pull/2

Is this what you are looking for?

**Reply by Koshdukai** [reply]
I'll check. Thanks !

**Reply by Koshdukai** [reply]
I wasn't able to build MiniDexed_Midi_Master but from what's described and looking at https://github.com/BenZonneveld/MiniDexed_Midi_Master/blob/master/MiniDexed_Master/UAC2/usb_audio.cpp I would say yes, that's what I'm talking about i.e. suggesting for MiniDexed, to support audio output through a class compliant USB audio interface :)

**Comment by probonopd** [comment]
Breaking news:

As per https://github.com/rsta2/circle/issues/222, there is now some initial support for USB Audio class devices on branch `usb-audio` of [circle](https://github.com/rsta2/circle). Thanks @rsta2. 

It looks like in order to use it, we'd have to
* Switch from the `develop` to the `usb-audio` branch of circle (probably this will be merged at some point)
* Backport the changes from miniorgan in https://github.com/rsta2/circle/commit/9ce864f0cc6a2b57b396f4067df3724315f5842d to MiniDexed

Any volunteers?

**Comment by probonopd** [comment]
Circle 45 now has support for USB audio devices on Raspberry Pi 4, 400, and CM4.

https://github.com/rsta2/circle/releases/tag/Step45

Essentially this would need to be ported over:

https://github.com/rsta2/circle/commit/b39206ab4143462c90f717aa02e817290f0472f4

Any volunteers?

**Reply by paolofalcoruegg** [reply]
Hey! Has this ever happened per chance? I would like to help, but unfortunately I‘m not familiar with the technologies.

**Reply by diyelectromusic** [reply]
Not as far as I'm aware, but I've made a note.  We need to fix HDMI audio on a RPi 5 too.

Kevin

**Comment by rsta2** [comment]
Adding USB audio support is not difficult. This small patch should do it (not tested):

```
diff --git a/src/minidexed.cpp b/src/minidexed.cpp
index da9e598..4452660 100644
--- a/src/minidexed.cpp
+++ b/src/minidexed.cpp
@@ -23,6 +23,7 @@
 #include <circle/sound/pwmsoundbasedevice.h>
 #include <circle/sound/i2ssoundbasedevice.h>
 #include <circle/sound/hdmisoundbasedevice.h>
+#include <circle/sound/usbsoundbasedevice.h>
 #include <circle/gpiopin.h>
 #include <string.h>
 #include <stdio.h>
@@ -186,6 +187,16 @@ CMiniDexed::CMiniDexed (CConfig *pConfig, CInterruptSystem *pInterrupt,
 		// The channels are swapped by default in the HDMI sound driver.
 		// TODO: Remove this line, when this has been fixed in the driver.
 		m_bChannelsSwapped = !m_bChannelsSwapped;
+#endif
+	}
+	else if (strcmp (pDeviceName, "usb") == 0)
+	{
+#if RASPPI<=3
+		LOGNOTE ("USB mode NOT supported on RPI 1-3.");
+#else
+		LOGNOTE ("USB mode");
+
+		m_pSoundDevice = new CUSBSoundBaseDevice (pConfig->GetSampleRate ());
 #endif
 	}
 	else
```

Please note that the driver by default searches for 16-bit audio interfaces. If your USB audio interfaces uses 24 bits, you have to add the option `soundopt=24` to the file *cmdline.txt* on the SD card.

**Reply by probonopd** [reply]
Thank you very much @rsta2. Now we need people with USB sound cards/interfaces to test this: https://github.com/probonopd/MiniDexed/pull/708

**Reply by Koshdukai** [reply]
Planning to, hopefully this weekend !

Thanks @rsta2 and @probonopd :)

**Comment by Koshdukai** [comment]
Tried it and it's half-a-success :)
https://github.com/probonopd/MiniDexed/pull/708#issuecomment-2316047653

btw, why are the RPi2-3 excluded from USB mode? Asking because the reason why I was using MiniDexed was to give some musical use to an old RPi2B for 1x DX7 ...because on RPi4~5 I can always use Zynthian which can run Dexed anyway :)

**Reply by rsta2** [reply, marked as answer]
RPi 1-3 has a different USB host controller, where isochronous transfers, which are necessary for USB audio streaming, are much more time critical, compared to the xHCI controllers of RPi 4-5. I got USB audio streaming working in a way on RPi 1-3, but not, when the USB MIDI driver was active at the same time. That's why this feature has been disabled there.

**Reply by Koshdukai** [reply]
oh ok, that makes sense "unfortunately" eheh, thanks!
